waku/v2/protocol/waku_noise/noise.nim

```diff
@@ -62,6 +77,94 @@ type
     nonce: ChaChaPolyNonce
     ad: seq[byte]
 
+  #################################
+  # Noise handshake patterns
+  #################################
+
+  # The Noise tokens appearing in Noise (pre)message patterns 
+  # as in http://www.noiseprotocol.org/noise.html#handshake-pattern-basics
+  NoiseTokens = enum
+    T_e = "e"
+    T_s = "s"
+    T_es = "es"
+    T_ee = "ee"
+    T_se = "se"
+    T_ss = "se"
+    T_psk = "psk"
+
+  # The direction of a (pre)message pattern in canonical form (i.e. Alice-initiated form)
+  # as in http://www.noiseprotocol.org/noise.html#alice-and-bob
+  MessageDirection* = enum
+    D_r = "->"
+    D_l = "<-"
+
+  # The pre message pattern consisting of a message direction and some Noise tokens, if any.
+  # (if non empty, only tokens e and s are allowed: http://www.noiseprotocol.org/noise.html#handshake-pattern-basics)
+  PreMessagePattern* = object
+    direction: MessageDirection
+    tokens: seq[NoiseTokens]
+
+  # The message pattern consisting of a message direction and some Noise tokens
+  # All Noise tokens are allowed
+  MessagePattern* = object
```

Comment:
They are independent since `PreMessagePattern` [can only](http://www.noiseprotocol.org/noise.html#handshake-patterns) have patterns`e`, `s`, `e,s` or none. However, the token processing is completely similar among `PreMessagePattern` and `MessagePattern` , so I kept their internal structure (that is, direction + token) the same to reuse code (in followup PR).

In this way, "validity" of a `PreMessagePattern` can be checked with a external utility function, rather than complicating data structures.

Reply:
Is there a validator implemented for the `PreMessagePattern`? if not, please add a TODO in the code to as not to forget to implement, also please add the constraint you explained here i.e., `PreMessagePattern can only have patterns, s, e,s or none` as a comment for the `PreMessagePattern` data type.

Reply:
By being related to handshake processing rules, such validator is already planned to be added in follow PR https://github.com/status-im/nwaku/pull/934.

Reply:
Added in https://github.com/status-im/nwaku/pull/934/commits/59857903fc4ffed4a345fc40459fec1d38b3798a